tests/unit/test_utils.py

```diff
@@ -203,3 +210,155 @@ def test_can_silence_io_error_from_pager(self):
                     pass
         except IOError:
             self.fail('Should not raise IOError')
+
+
+class TestInstanceMetadataRegionFetcher(unittest.TestCase):
+    def setUp(self):
+        urllib3_session_send = 'botocore.httpsession.URLLib3Session.send'
+        self._urllib3_patch = mock.patch(urllib3_session_send)
+        self._send = self._urllib3_patch.start()
+        self._imds_responses = []
+        self._send.side_effect = self.get_imds_response
+        self._region = 'us-mars-1a'
+
+    def tearDown(self):
+        self._urllib3_patch.stop()
+
+    def add_imds_response(self, body, status_code=200):
+        response = botocore.awsrequest.AWSResponse(
+            url='http://169.254.169.254/',
+            status_code=status_code,
+            headers={},
+            raw=RawResponse(body)
+        )
+        self._imds_responses.append(response)
+
+    def add_get_region_imds_response(self, region=None):
+        if region is None:
+            region = self._region
+        self.add_imds_response(body=region.encode('utf-8'))
+
+    def add_imds_connection_error(self, exception):
+        self._imds_responses.append(exception)
+
+    def get_imds_response(self, request):
+        response = self._imds_responses.pop(0)
+        if isinstance(response, Exception):
+            raise response
+        return response
+
+    def test_disabled_by_environment(self):
+        env = {'AWS_EC2_METADATA_DISABLED': 'true'}
+        fetcher = InstanceMetadataRegionFetcher(env=env)
+        result = fetcher.retrieve_region()
+        self.assertIsNone(result)
+        self._send.assert_not_called()
+
+    def test_disabled_by_environment_mixed_case(self):
+        env = {'AWS_EC2_METADATA_DISABLED': 'tRuE'}
+        fetcher = InstanceMetadataRegionFetcher(env=env)
+        result = fetcher.retrieve_region()
+        self.assertIsNone(result)
+        self._send.assert_not_called()
+
+    def test_disabling_env_var_not_true(self):
+        url = 'https://example.com/'
+        env = {'AWS_EC2_METADATA_DISABLED': 'false'}
+
+        self.add_get_region_imds_response()
+
+        fetcher = InstanceMetadataRegionFetcher(base_url=url, env=env)
+        result = fetcher.retrieve_region()
+
+        expected_result = 'us-mars-1'
+        self.assertEqual(result, expected_result)
+
+    def test_includes_user_agent_header(self):
+        user_agent = 'my-user-agent'
+        self.add_get_region_imds_response()
+
+        InstanceMetadataRegionFetcher(
+            user_agent=user_agent).retrieve_region()
+
+        headers = self._send.call_args[0][0].headers
+        self.assertEqual(headers['User-Agent'], user_agent)
+
+    def test_non_200_response_for_region_is_retried(self):
```

Comment:
Do we have a test for the happy 200 response case?

Reply:
Yea there is a test for when imds is not disabled.

https://github.com/aws/aws-cli/pull/3680/files/e7d666c14c84b774c8f8fa6bfe5b25118f05c962#diff-4b935d93f7e4f4a47577b93a23022fe3R264